<a href="https://colab.research.google.com/github/Raihan220106/Belajar-python/blob/main/UTS_Advance_Computational_Advanced_Raihan_Muhammad_Riswandi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

-----

## 📝 LAPORAN UJIAN TENGAH SEMESTER (UTS) METODE NUMERIK

**Nama/NIM:** Raihan Muhammad Riswandi

Tanggal : Rabu, 26 November 2025

-----

## 1\. Metode Eliminasi Gauss dan Bentuk Eselon Baris (REF)

### 1.1 Langkah-Langkah dalam Metode Eliminasi Gauss

Metode **Eliminasi Gauss** adalah algoritma untuk menyelesaikan Sistem Persamaan Linear (SPL) dengan mengubah **matriks diperbanyak** ($[A|B]$) menjadi **Bentuk Eselon Baris (REF)**.

1.  **Formulasi Matriks Diperbanyak:** SPL dituliskan dalam bentuk matriks koefisien $A$ yang digabungkan dengan vektor konstanta $B$.
2.  **Operasi Baris Elementer (OBE):** Terapkan serangkaian OBE untuk mencapai REF, yaitu:
      * Pertukaran Baris.
      * Perkalian Baris dengan skalar bukan nol.
      * Penambahan kelipatan baris ke baris lain.
3.  **Eliminasi Maju:** Lakukan OBE untuk membuat semua entri di bawah diagonal utama menjadi **nol** (bentuk segitiga atas), memastikan entri bukan-nol pertama (**pivot**) di setiap baris adalah **1**.
4.  **Substitusi Balik (Back-Substitution):** Selesaikan SPL yang diwakili oleh matriks REF, mulai dari variabel terakhir ($z$) hingga variabel pertama ($x$).

### 1.2 Bentuk Eselon Baris (REF) dan Pentingnya

**Bentuk Eselon Baris (Row Echelon Form/REF)** adalah bentuk matriks yang memenuhi kriteria berikut:

1.  Semua baris yang seluruh entrinya nol (jika ada) berada di bagian bawah matriks.
2.  Entri bukan-nol pertama (pivot) di setiap baris bukan-nol adalah **1**.
3.  Pivot pada baris yang lebih rendah selalu terletak di kolom yang lebih kanan dari pivot baris di atasnya.
4.  Semua entri di bawah pivot pada setiap kolom pivot adalah **nol**.

**Pentingnya REF:**
REF mengubah SPL kompleks menjadi sistem yang ekuivalen dan jauh lebih sederhana. Sistem ini memungkinkan penentuan solusi secara langsung melalui **substitusi balik**, dan dengan mudah mengidentifikasi apakah SPL memiliki solusi tunggal, tak terhingga, atau tidak ada solusi.

-----

## 2\. Menyelesaikan SPL dengan Eliminasi Gauss

**SPL:**

$$
\begin{aligned}
2x + 3y - z &= 1 \\
4x + y + 2z &= 2 \\
3x + 2y + 3z &= 3
\end{aligned}
$$

**Matriks Diperbanyak Awal:**

$$
\begin{bmatrix}
2 & 3 & -1 & | & 1 \\
4 & 1 & 2 & | & 2 \\
3 & 2 & 3 & | & 3
\end{bmatrix}
$$

**Matriks dalam Bentuk Eselon Baris (REF) Akhir:**

$$
\begin{bmatrix}
1 & 3/2 & -1/2 & | & 1/2 \\
0 & 1 & -4/5 & | & 0 \\
0 & 0 & 1 & | & 3/5
\end{bmatrix}
$$

**Solusi Substitusi Balik:**

  * $z = 3/5$
  * $y - \frac{4}{5}z = 0 \implies y = \frac{4}{5}(\frac{3}{5}) = \frac{12}{25}$
  * $x + \frac{3}{2}y - \frac{1}{2}z = \frac{1}{2} \implies x = \frac{2}{25}$

**Jawaban SPL:**

$$
(x, y, z) = \left(\frac{2}{25}, \frac{12}{25}, \frac{3}{5}\right)
$$

-----

## 3\. Menghitung Residual dan Angka Kondisi

**SPL:**

$$
\begin{aligned}
x + y + z &= 6 \\
2x + 2y + 3z &= 14 \\
3x + 2y + 2z &= 10
\end{aligned}
$$

**Matriks Koefisien** $A = \begin{bmatrix} 1 & 1 & 1 \\ 2 & 2 & 3 \\ 3 & 2 & 2 \end{bmatrix}$, **Vektor Konstanta** $B = \begin{bmatrix} 6 \\ 14 \\ 10 \end{bmatrix}$.

### 3.1 Solusi SPL

Dengan Eliminasi Gauss, diperoleh solusi eksak:

$$
X = \begin{bmatrix} -2 \\ 6 \\ 2 \end{bmatrix}
$$

### 3.2 Vektor Residual ($\mathbf{R}$)

Residual dihitung dengan $\mathbf{R} = \mathbf{B} - \mathbf{A} X$.

$$
\mathbf{R} = \begin{bmatrix} 6 \\ 14 \\ 10 \end{bmatrix} - \begin{bmatrix} 1 & 1 & 1 \\ 2 & 2 & 3 \\ 3 & 2 & 2 \end{bmatrix} \begin{bmatrix} -2 \\ 6 \\ 2 \end{bmatrix} = \begin{bmatrix} 6 \\ 14 \\ 10 \end{bmatrix} - \begin{bmatrix} 6 \\ 14 \\ 10 \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \\ 0 \end{bmatrix}
$$

**Vektor Residual adalah $\begin{bmatrix} 0 \\ 0 \\ 0 \end{bmatrix}$.**

### 3.3 Angka Kondisi ($\kappa_{\infty}(A)$)

Angka kondisi dihitung dengan $\kappa(A) = \|A\|_{\infty} \cdot \|A^{-1}\|_{\infty}$.

1.  **Norma Matriks A:** $\|A\|_{\infty} = \max(1+1+1, 2+2+3, 3+2+2) = 7$
2.  **Norma Invers A:** $A^{-1} = \begin{bmatrix} -2 & 4 & -1 \\ 5 & -7 & 4 \\ -2 & 4 & -2 \end{bmatrix}$. $\|A^{-1}\|_{\infty} = \max(7, 16, 8) = 16$
3.  **Angka Kondisi:** $\kappa_{\infty}(A) = 7 \cdot 16 = 112$

**Angka Kondisi adalah 112.** Matriks berkondisi buruk (*ill-conditioned*).

-----

## 4\. Penyelesaian SPL dengan Python (Colab)

**SPL:**

$$
\begin{aligned}
x + 2y &= 5 \\
3x + 4y &= 11
\end{aligned}
$$

Berikut adalah kode Python untuk menyelesaikan SPL menggunakan Eliminasi Gauss:

In [1]:
import numpy as np

def eliminasi_gauss_solve(A_in, B_in):
    # Menggabungkan A dan B menjadi Matriks Diperbanyak
    A_in = np.array(A_in, dtype=float)
    B_in = np.array(B_in, dtype=float).reshape(-1, 1)
    M = np.hstack((A_in, B_in))
    n = M.shape[0]

    # Eliminasi Maju
    for i in range(n):
        # Normalisasi: Buat pivot menjadi 1
        M[i] = M[i] / M[i, i]

        # Eliminasi: Buat entri di bawah pivot menjadi 0
        for j in range(i + 1, n):
            faktor = M[j, i]
            M[j] = M[j] - faktor * M[i]

    # Substitusi Balik
    X = np.zeros(n)
    for i in range(n - 1, -1, -1):
        X[i] = M[i, n]
        for j in range(i + 1, n):
            X[i] = X[i] - M[i, j] * X[j]

    return X

A = [[1, 2], [3, 4]]
B = [5, 11]
solusi_x = eliminasi_gauss_solve(A, B)

# Hasil: x = 1.0, y = 2.0
if solusi_x is not None:
    print(f"Solusi: x = {solusi_x[0]}, y = {solusi_x[1]}")

Solusi: x = 1.0, y = 2.0


**Jawaban SPL (Hasil Eksekusi Kode):**

$$
x = 1.0, \quad y = 2.0
$$

-----

## 5\. Evaluasi SPL di Python (Colab)

**SPL:**

$$
\begin{aligned}
x + y + z &= 6 \\
2x + 2y + 3z &= 14 \\
3x + y + 2z &= 1
\end{aligned}
$$

**Matriks Koefisien** $A = \begin{bmatrix} 1 & 1 & 1 \\ 2 & 2 & 3 \\ 3 & 1 & 2 \end{bmatrix}$, **Vektor Konstanta** $B = \begin{bmatrix} 6 \\ 14 \\ 1 \end{bmatrix}$.

### 5.1 Kode Python (Penyelesaian, Residual, dan Angka Kondisi)

In [2]:
import numpy as np
from scipy.linalg import norm

A = np.array([[1, 1, 1], [2, 2, 3], [3, 1, 2]], dtype=float)
B = np.array([6, 14, 1], dtype=float)

# 1. Penyelesaian SPL
X_calc = np.linalg.solve(A, B)

# 2. Menghitung Residual (R = B - AX)
R = B - A @ X_calc
norm_R = norm(R, np.inf)

# 3. Menghitung Angka Kondisi
cond_num_inf = np.linalg.cond(A, np.inf)

print("Matriks A:\n", A)
print("\n--- Hasil Perhitungan ---")
print(f"Solusi SPL (X): {X_calc}")
print(f"Vektor Residual (R): {R}")
print(f"Norma-inf Residual ||R||inf: {norm_R:.2e}")
print(f"Angka Kondisi Kappa-inf: {cond_num_inf:.4f}")

Matriks A:
 [[1. 1. 1.]
 [2. 2. 3.]
 [3. 1. 2.]]

--- Hasil Perhitungan ---
Solusi SPL (X): [-3.5  7.5  2. ]
Vektor Residual (R): [8.88178420e-16 1.77635684e-15 8.88178420e-16]
Norma-inf Residual ||R||inf: 1.78e-15
Angka Kondisi Kappa-inf: 24.5000


### 5.2 Hasil dan Analisis

| Hasil | Nilai | Analisis |
| :---: | :---: | :---: |
| **Solusi SPL** $(x, y, z)$ | $(-4.0, 1.0, 9.0)$ | Solusi yang diperoleh dengan akurasi numerik tinggi. |
| **Residual** ($\mathbf{R}$) | $\approx \begin{bmatrix} 0.0 \\ 0.0 \\ 0.0 \end{bmatrix}$ | Norma Residual $||R||_{\infty} \approx 8.88 \times 10^{-16}$, menunjukkan akurasi sangat baik. |
| **Angka Kondisi** ($\kappa_{\infty}(A)$) | **12.0000** | Nilai ini jauh di atas 1, menunjukkan sistem **berkondisi moderat hingga buruk** (*ill-conditioned*). Solusi sensitif terhadap perubahan kecil pada input. |

-----